src/clj/web/user.clj

```diff
@@ -9,6 +9,28 @@
    :isadmin :ismoderator :tournament-organizer
    :special :options :stats :has-api-keys :banned])
 
+(def max-username-chars 20)
+(def ascii-invalid-char-pattern #"[^\w.-]")
+(def unicode-invalid-char-pattern #"[^\p{Alnum}\p{M}._-]")
+(def invalid-leading-char-pattern #"^[^\p{Alnum}\p{M}_]+")
+(def invalid-trailing-char-pattern #"[^\p{Alnum}\p{M}]+$")
+(def repeated-special-char-pattern #"[-_.]{2,}")
+(def patterns-username [
+                        ascii-invalid-char-pattern,
+                        unicode-invalid-char-pattern,
+                        invalid-leading-char-pattern,
+                        invalid-trailing-char-pattern,
+                        repeated-special-char-pattern
+                        ])
+
+(defn valid-username?
+  "Validate a username"
+  [username]
+  (if (<= (count username) max-username-chars)
+    (every? (fn [re] (= (re-matches re username) false)) patterns-username)
```

Comment:
(= X false) can be (not X). But also, (every? (not X) Y) should be (not (any? X Y)). So:

```suggestion
    (not (any? #(re-matches % username) patterns-username))
```
I'm on my phone so test this first but it should be equivalent while also short-circuiting on the first match.

Reply:
Heh, I was going to use function literal shorthand here but was worried it was unconventional.

I see you prefer to compose functions vs using something like every - sounds good to me :slightly_smiling_face: 

I'll look around the codebase to see where to add tests...